cmd/endpoint.go

```diff
@@ -0,0 +1,425 @@
+			// As serverAddr is given, serverAddr and endpoint should have same port.
+			if serverAddrPort != port {
+				return serverAddr, endpoints, setupType, fmt.Errorf("server address and endpoint have different ports")
+			}
+		}
+
+		endpoints, _ = NewEndpointList(normArgs...)
+		setupType = XLSetupType
+		return serverAddr, endpoints, setupType, nil
+	}
+
+	isXLSetup := true
+	localhostPort := ""
+	var uniqueLocalHosts []string
+	{
+		uniqueLocalHostSet := set.NewStringSet()
+		for i, endpoint := range endpoints {
+			if endpoint.IsLocal {
+				uniqueLocalHostSet.Add(endpoint.Host)
+
+				_, port := mustSplitHostPort(endpoint.Host)
+				if i == 0 {
+					localhostPort = port
+				} else if localhostPort != port {
+					// Eventhough this endpoint is for local host, but previously
+					// assigned port of localhost and this endpoint's port is different.
+					// This means that not all endpoints are local ie not a XL setup.
+					isXLSetup = false
+				}
+			} else {
+				isXLSetup = false
+			}
+		}
+
+		// Error out if no endpoint for this server.
+		if uniqueLocalHostSet.IsEmpty() {
+			return serverAddr, endpoints, setupType, fmt.Errorf("no endpoint found for this host")
+		}
+
+		uniqueLocalHosts = uniqueLocalHostSet.ToSlice()
+	}
+
+	if isXLSetup {
+		// TODO: In this case, we bind to 0.0.0.0 ie to all interfaces.
+		// The actual way to do is bind to only IPs in uniqueLocalHosts.
+		endpoints, _ = NewEndpointList(normArgs...)
+		serverAddr = net.JoinHostPort("", localhostPort)
+		setupType = XLSetupType
+		return serverAddr, endpoints, setupType, nil
+	}
+
+	// This is Distribute setup.
+	if len(uniqueLocalHosts) == 1 {
+		host, port := mustSplitHostPort(uniqueLocalHosts[0])
+		if isServerAddrEmpty {
+			serverAddr = net.JoinHostPort(host, port)
+		} else {
+			// As serverAddr is given, serverAddr and endpoint should have same port.
+			if serverAddrPort != port {
+				return serverAddr, endpoints, setupType, fmt.Errorf("server address and endpoint have different ports")
+			}
+		}
+	} else {
+		// If length of uniqueLocalHosts is more than one,
+		// server address should be present with same port with the same or empty hostname.
+		if isServerAddrEmpty {
+			return serverAddr, endpoints, setupType, fmt.Errorf("for more than one endpoints for local host with different port, server address must be provided")
+		}
+
+		// Check if port in server address matches to port in at least one unique local hosts.
+		found := false
+		for _, host := range uniqueLocalHosts {
+			_, port := mustSplitHostPort(host)
+			if serverAddrPort == port {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return serverAddr, endpoints, setupType, fmt.Errorf("port in server address does not match with local endpoints")
+		}
+	}
+
+	setupType = DistXLSetupType
+	return serverAddr, endpoints, setupType, nil
+}
+
+// GetRemotePeers - get hosts information other than this minio service.
+func GetRemotePeers(endpoints EndpointList) []string {
```

Comment:
It was in previous refresh. As its using globals, its not a good idea to be in a type.